plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java

```diff
@@ -439,8 +439,13 @@ private static PreparedQuery asPreparedQuery(JdbcTableHandle tableHandle)
             return Optional.empty();
         }
 
+        boolean limitGuaranteed = jdbcClient.isLimitGuaranteed(session);
+
         if (handle.getLimit().isPresent() && handle.getLimit().getAsLong() <= limit) {
-            return Optional.empty();
+            if (!limitGuaranteed) {
+                return Optional.empty();
+            }
+            return Optional.of(new LimitApplicationResult<>(handle, limitGuaranteed, false));
```

Comment:
connector should return non-`empty` only when plan changes.
per `io.trino.spi.connector.ConnectorMetadata#applyLimit` doc:

```
     * Connectors can indicate whether they don't support limit pushdown or that the action had no effect
     * by returning {@link Optional#empty()}. 
```

Reply:
I think instead, we would need "redundant limit pruning rule"
i.e. a connector needs to be able to tell the engine what's the max cardinality to be returned
(just like it can return enforced predicates)